### PR TITLE
Fix potential deadlock

### PR DESCRIPTION
**Pull request by MBkkt**
[Unbenannt.txt](https://github.com/iresearch-toolkit/iresearch/files/7721123/Unbenannt.txt)

### CODE REVIEW

File: core/utils/async_utils.cpp [hunk 1]
```diff
@@ -327,56 +327,38 @@ void thread_pool::worker_impl(std::unique_lock<std::mutex>& lock,
 
   while (State::ABORT != state.load() && threads_.load() <= max_threads_) {
     if (!queue_.empty()) {
-      auto& top = queue_.top();
```
Comment (MBkkt) [inline core/utils/async_utils.cpp:330]: It is const ref, because priority queue

File: core/utils/async_utils.cpp [hunk 2]
```diff
@@ -327,56 +327,38 @@ void thread_pool::worker_impl(std::unique_lock<std::mutex>& lock,
...
-
-      if (top.at <= clock_t::now()) {
+      if (auto const& top = queue_.top(); top.at <= clock_t::now()) {
         func_t fn;
+        fn.swap(const_cast<std::function<void()>&>(top.fn));
```
Comment (MBkkt) [inline core/utils/async_utils.cpp:332]: swap noexcept

File: core/utils/async_utils.cpp [hunk 3]
```diff
@@ -327,56 +327,38 @@ void thread_pool::worker_impl(std::unique_lock<std::mutex>& lock,
...
-          } catch (const std::exception& e) {
-            IR_FRMT_ERROR("Failed to execute task, error '%s'", e.what());
-          } catch (...) {
-            IR_FRMT_ERROR("Failed to execute task");
-          }
-
-          lock.lock();
+        try {
+          fn();
+          fn = [] {};
```
Comment (MBkkt) [inline core/utils/async_utils.cpp:350]: Call dtor before mutex lock

**Review by Dronplane — APPROVED** [review]
LGTM

**Review by alexbakharew — APPROVED** [review]

**Review by mpoeter — APPROVED** [review]

File: core/utils/async_utils.cpp (hunk 2 above)
Comment (gnusi) [reply core/utils/async_utils.cpp:333]: good point!

File: core/utils/async_utils.cpp [hunk 4]
```diff
@@ -326,62 +326,41 @@ void thread_pool::worker_impl(std::unique_lock<std::mutex>& lock,
   lock.lock();
 
   while (State::ABORT != state.load() && threads_.load() <= max_threads_) {
+    assert(lock.owns_lock());
     if (!queue_.empty()) {
-      auto& top = queue_.top();
-
-      if (top.at <= clock_t::now()) {
+      if (auto const& top = queue_.top(); top.at <= clock_t::now()) {
```
Comment (gnusi) [inline core/utils/async_utils.cpp:331]: `auto const&` -> `const auto&`

File: core/utils/async_utils.cpp (hunk 3 above)
Comment (gnusi) [reply core/utils/async_utils.cpp:352]: nice catch! but maybe just `fn = {}` ?

File: core/utils/async_utils.cpp (hunk 4 above)
Comment (MBkkt) [reply core/utils/async_utils.cpp]: Done

File: core/utils/async_utils.cpp (hunk 3 above)
Comment (MBkkt) [reply core/utils/async_utils.cpp]: done

**Pull request merged**